LibreHardwareMonitorLib/LibreHardwareMonitorLib.csproj

```diff
@@ -40,4 +40,9 @@
     <PackageReference Include="HidSharp" Version="2.1.0" />
     <PackageReference Include="System.Management" Version="4.7.0" />
   </ItemGroup>
+  <ItemGroup>
+    <None Update="inpoutx64.dll">
+      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
+    </None>
+  </ItemGroup>
```

Comment:
What is the source, and license, of this DLL?

Reply:
https://www.highrez.co.uk/Downloads/InpOut32/

Reply:
This driver doesn't seem maintained anymore, but perhaps more importantly:

>  There was, for a time, an input output project called **WinRing0**. It had its own signed 64bit drivers and was open source.

WinRing0 is already used by LHM, can you change this to use that instead? That would save an additional driver.

Reply:
I could not get the included WinRing0 to work.

Maybe not compiled with `_PHYSICAL_MEMORY_SUPPORT` enabled?

Reply:
Have you tried another build of WinRing0? There are some other precompiled versions floating around.

Reply:
Yes, I have tried; all are the same build as included here.

Should I remove this unsafe dll and the Ryzen SMU support from the pull request?

Reply:
Here's a newer version (on official site): https://openlibsys.org/
It's included in VirtualChecker zip. I'd rather avoid loading another driver, I've checked InpOut32 and I don't think it's possible to replace WinRing0 with that (correct me if I'm wrong).

Reply:
No luck with the version in VirtualChecker also.
You are correct about that it's not possible to replace WinRing0 with InpOut32.

This is how I try to read memory with WinRing0:
```csharp
        // in RyzenSMU.cs
        private float[] ReadDramToArray()
        {
            float[] table = new float[_pm_table_size / 4];

            for (uint i = 0; i < table.Length; i++)
            {
                Ring0.ReadMemory(_dram_base_addr.ToUInt64(), ref table[i]);
            }

            return table;
        }
```
I think it should work, yet it returns garbage.

Reply:
I think it should be possible in Ring0. I think you need to offset `_dram_base_addr` with the table index, otherwise you're only reading the first address. Another thing, is InpOut32 perhaps doing some address translation and WinRing0 is reading the wrong address?

Reply:
Relevant code:
  https://github.com/ellysh/InpOut32/blob/master/driver/hwinterfacedrv.c#L280
vs
  https://github.com/QCute/WinRing0/blob/master/dll/sys/OpenLibSys.c#L603

I can't find the difference in the addressing.

Reply:
I can't test it properly on my machine due to my CPU not being supported, butI think the parameters to `ReadMemory` might be wrong, as invalid parameter is returned by the function. Until a solution to this has been found, such as a working version of WinRing0, I think it's best to put the SMU bit on hold.

Reply:
Does https://github.com/irusanov/SMUDebugTool works with your CPU?

Reply:
Yes, seems to work.

Reply:
I have it working with another WinRing0 version (values are identical to InpOut32, whatever they mean):

![image](https://user-images.githubusercontent.com/7643972/114303142-defc1380-9acc-11eb-9fb0-3ee60880e2b8.png)

[WinRing0.zip](https://github.com/LibreHardwareMonitor/LibreHardwareMonitor/files/6291667/WinRing0.zip)

Reply:
Nice!

I used this and hwinfo to match values for my cpu:
https://github.com/moson-mo/rpms/blob/main/tables.go

How should we move forward?